src/libsyntax/feature_gate.rs

```diff
@@ -1220,72 +1221,114 @@ pub const BUILTIN_ATTRIBUTES: &[(&str, AttributeType, AttributeTemplate, Attribu
-                               "#[optimize] attribute is an unstable feature",
-                               cfg_fn!(optimize_attribute))),
+    (optimize, Whitelisted, template!(List: "size|speed"), Gated(Stability::Unstable,
+                            "optimize_attribute",
+                            "#[optimize] attribute is an unstable feature",
+                            cfg_fn!(optimize_attribute))),
 
     // Crate level attributes
-    ("crate_name", CrateLevel, template!(NameValueStr: "name"), Ungated),
-    ("crate_type", CrateLevel, template!(NameValueStr: "bin|lib|..."), Ungated),
-    ("crate_id", CrateLevel, template!(NameValueStr: "ignored"), Ungated),
-    ("feature", CrateLevel, template!(List: "name1, name1, ..."), Ungated),
-    ("no_start", CrateLevel, template!(Word), Ungated),
-    ("no_main", CrateLevel, template!(Word), Ungated),
-    ("no_builtins", CrateLevel, template!(Word), Ungated),
-    ("recursion_limit", CrateLevel, template!(NameValueStr: "N"), Ungated),
-    ("type_length_limit", CrateLevel, template!(NameValueStr: "N"), Ungated),
-    ("test_runner", CrateLevel, template!(List: "path"), Gated(Stability::Unstable,
+    (crate_name, CrateLevel, template!(NameValueStr: "name"), Ungated),
+    (crate_type, CrateLevel, template!(NameValueStr: "bin|lib|..."), Ungated),
+    (crate_id, CrateLevel, template!(NameValueStr: "ignored"), Ungated),
+    (feature, CrateLevel, template!(List: "name1, name1, ..."), Ungated),
+    (no_start, CrateLevel, template!(Word), Ungated),
+    (no_main, CrateLevel, template!(Word), Ungated),
+    // FIXME: Duplicated
+    //(no_builtins, CrateLevel, template!(Word), Ungated),
+    (recursion_limit, CrateLevel, template!(NameValueStr: "N"), Ungated),
+    (type_length_limit, CrateLevel, template!(NameValueStr: "N"), Ungated),
+    (test_runner, CrateLevel, template!(List: "path"), Gated(Stability::Unstable,
                     "custom_test_frameworks",
                     EXPLAIN_CUSTOM_TEST_FRAMEWORKS,
                     cfg_fn!(custom_test_frameworks))),
-];
+}}}
+
+type BuiltinAttribute = (Symbol, AttributeType, AttributeTemplate, AttributeGate);
+
+macro_rules! declare_builtin_attributes {
+    ($(($name:ident, $level:expr, $template: expr, $gate: expr),)*) => {
+        /// Attributes that have a special meaning to rustc or rustdoc
+        pub const BUILTIN_ATTRIBUTES: &[BuiltinAttribute] = &[
+            $((symbols::$name, $level, $template, $gate)),*
+        ];
+    }
+}
+
+macro_rules! find_builtin_attribute_int {
+    ($(($num:expr, ($name:ident, $level:expr, $template: expr, $gate: expr)))*) => {
+        pub fn find_builtin_attr(name: Symbol) -> Option<&'static BuiltinAttribute> {
+            // Check for duplicate attributes
+            $(#[allow(unused_imports)] use symbols::$name;)*
+
+            debug_assert!(!name.is_gensymed());
+            match name {
+                $(
+                    symbols::$name => Some(&BUILTIN_ATTRIBUTES[$num]),
```

Comment:
I wonder how this huge match is codegen-ed and whether it's better than linear search or not.
The number of attributes seems small enough that we could avoid it and keep scanning the attribute array linearly.

Reply:
116 * `u32` = ~8 consecutive cache lines (if the data is split into two arrays), that's pretty good for linear search.
(And my primary goal is to avoid stressing the people and the compiler with all those extra macros and huge matches, of course.)

Reply:
Collecting them into a hash table might be good enough. If the attributes had the same order in the `symbol.rs` a match is probably best though.

Reply:
LLVM creates a jump table with 121 cases, not ideal but decent.

Reply:
I've changed the code to use a hash map here, no marcos used. Seems to perform just as well.